colabs/wandb_registry/zoo_wandb.ipynb

```diff
@@ -0,0 +1,984 @@
+    "f\"{ORG_NAME}/wandb-registry-{REGISTRY_NAME}/{COLLECTION_NAME}\"\n",
+    "\n",
+    "```\n",
+    "\n",
+    "```python\n",
+    "# Artifact name/filepath for downloading and using artifact publsihed in a registry\n",
+    "f\"{ORG_NAME}/wandb-registry-{REGISTRY_NAME}/{COLLECTION_NAME}:v{VERSION}\"\n",
+    "```\n",
+    "\n",
+    "Since we only linked on artifact version, the version we'll use is `v0`. (W&B uses 0 indexing)."
+   ]
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "id": "1042375b-c2c8-4503-9888-a406bc9c5df1",
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "run = wandb.init(entity = TEAM_ENTITY, project = PROJECT, job_type = \"training\", config = hyperparameter_config)\n",
+    "\n",
+    "# Get dataset artifacts from registry\n",
+    "VERSION = 0\n",
+    "artifact_name = f\"{ORG_NAME}/wandb-registry-{REGISTRY_NAME.lower()}/{COLLECTION_NAME}:v{VERSION}\"\n",
+    "dataset_artifact = run.use_artifact(artifact_or_name=artifact_name)\n",
+    "\n",
+    "# Download only the training data\n",
+    "X_train_path = dataset_artifact.download(path_prefix=\"zoo_dataset_X_train\")\n",
+    "y_train_path = dataset_artifact.download(path_prefix=\"zoo_labels_y_train\")\n",
+    "\n",
+    "# Load data as tensors \n",
+    "X_train = torch.load(f=X_train_path+\"/zoo_dataset_X_train\")\n",
+    "y_train = torch.load(f=y_train_path+\"/zoo_labels_y_train\")\n",
+    "\n",
+    "# Set initial dummy loss value to compare to in training loop\n",
+    "prev_best_loss = 1e10 \n",
+    "\n",
+    "# Keep track of which model version for us to use at a later step\n",
+    "# Set to -1 for 0 indexing\n",
+    "model_version = -1\n",
+    "\n",
+    "# Training loop\n",
+    "for e in range(hyperparameter_config[\"epochs\"] + 1):\n",
+    "    pred = model(X_train)\n",
+    "    loss = loss_fn(pred, y_train.squeeze(1))\n",
+    "    \n",
+    "    loss.backward()\n",
+    "    optimizer.step()\n",
+    "    optimizer.zero_grad()\n",
+    "\n",
+    "    wandb.log({\n",
+    "            \"train/epoch_ndx\": e,\n",
+    "            \"train/train_loss\": loss\n",
+    "        })\n",
+    "\n",
+    "    # Checkpoint/save model if loss improves\n",
+    "    if (e % 100 == 0) and (loss <= prev_best_loss):\n",
+    "        print(\"epoch: \", e, \"loss:\", loss.item())\n",
+    "    \n",
+    "        PATH = 'zoo_wandb.pth' \n",
+    "        torch.save(model.state_dict(), PATH)\n",
+    "\n",
+    "        model_artifact_name = f\"zoo-{wandb.run.id}\"\n",
+    "        artifact = wandb.Artifact(\n",
+    "            name=model_artifact_name,\n",
+    "            type=\"model\",\n",
+    "            metadata={\n",
+    "                \"num_classes\": 7,\n",
+    "                \"model_type\": wandb.config[\"model_type\"]\n",
+    "            }\n",
+    "        )\n",
+    "        \n",
+    "        # Add artifact file\n",
+    "        artifact.add_file(PATH)\n",
```

Comment:
Shouldn't this be outside the for loop so we only save the best model? just save a ref to the best model inside the for loop, otherwise we are saving a bunch of iterations that aren't necessary, or are we doing this on purpose to showcase versioning?

Reply:
Yeah, the idea was to show versioning. But, thinking about it more, showing it this way might convolute the example.

In a real world setting, I suspect versioning is more intuitive to see, use as they are messing about with hyperparameters, modifying the neural network itself, etc. 

TL;DR Showing versioning here as a byproduct of checkpointing might not be intuitive, useful for showing why, how to use versioning.